jupyter_client/multikernelmanager.py

```diff
@@ -260,10 +260,11 @@ async def _async_shutdown_kernel(
         if self._using_pending_kernels() and kernel_id in self._pending_kernels:
             raise RuntimeError("Kernel is in a pending state. Cannot shutdown.")
         # If the kernel is still starting, wait for it to be ready.
-        elif kernel_id in self._starting_kernels:
-            kernel = self._starting_kernels[kernel_id]
+        elif kernel_id in self._pending_kernels:
+            kernel = self._pending_kernels[kernel_id]
             try:
                 await kernel
+                await kernel.ready
```

Comment:
I think I see the issue—you are on the right track. The naming of things is getting a little messy.

In this line, `kernel` is not the kernel manager, like you'd expect. It's the `Future` returned by `_add_kernel_when_ready`. Waiting this `Future` doesn't ensure that kernel has started, it just ensures that the `start_kernel` method returned. When pending kernels are enabled, `start_kernel` returns immediately, but the kernel might not be ready. To check that the kernel is ready, we need to call `await ... .ready` on the actual kernel manager.


Calling `.ready` on this future is redundant. It might be falsely helping the JupyterLab tests pass because it slows the method enough to dodge the race condition.